src/UglyToad.PdfPig/Tokenization/Scanner/PdfTokenScanner.cs

```diff
@@ -330,150 +331,129 @@ private bool TryReadStream(long startStreamTokenOffset, bool getLength, [NotNull
-
-                    read++;
                 }
-
-                binaryWrite.Flush();
-
-                if (observedEndLocations.Count == 0)
+                else
                 {
-                    return false;
+                    // For safety reset every counter in case we had a partial read.
+
+                    endStreamPosition = 0;
+                    endObjPosition = 0;
+                    commonPartPosition = (inputBytes.CurrentByte == endWordPart[0]) ? 1 : 0;
                 }
 
-                memoryStream.Seek(0, SeekOrigin.Begin);
-                if (length.HasValue && memoryStream.Length >= length)
-                {
-                    // Use the declared length to copy just the data we want.
-                    byte[] data = new byte[length.Value];
+                read++;
+            }
 
-                    memoryStream.Read(data, 0, (int)length.Value);
+            long streamDataEnd = inputBytes.CurrentOffset + 1;
 
-                    stream = new StreamToken(streamDictionaryToken, data);
-                }
-                else
-                {
-                    // Work out where '\r\nendobj' or '\r\nendstream' occurs and read everything up to that.
-                    var lastEnd = observedEndLocations[observedEndLocations.Count - 1];
+            if (possibleEndLocation == null)
+                return false;
 
-                    var dataLength = lastEnd.Offset - startDataOffset;
+            var lastEnd = possibleEndLocation;
 
-                    var current = inputBytes.CurrentOffset;
+            var dataLength = lastEnd.Value.Offset - startDataOffset;
 
-                    // 3 characters, 'e', '\n' and possibly '\r'
-                    inputBytes.Seek(lastEnd.Offset - 3);
-                    inputBytes.MoveNext();
+            // 3 characters, 'e', '\n' and possibly '\r'
+            inputBytes.Seek(lastEnd.Value.Offset - 3);
+            inputBytes.MoveNext();
 
-                    if (inputBytes.CurrentByte == '\r')
-                    {
-                        dataLength -= 3;
-                    }
-                    else
-                    {
-                        dataLength -= 2;
-                    }
+            if (inputBytes.CurrentByte == '\r')
+            {
+                dataLength -= 3;
+            }
+            else
+            {
+                dataLength -= 2;
+            }
 
-                    inputBytes.Seek(current);
+            Span<byte> data = new byte[dataLength];
 
-                    byte[] data = new byte[dataLength];
+            inputBytes.Seek(streamDataStart);
+            inputBytes.Read(data);
 
-                    memoryStream.Read(data, 0, (int)dataLength);
+            inputBytes.Seek(streamDataEnd);
 
-                    stream = new StreamToken(streamDictionaryToken, data);
-                }
-            }
+            stream = new StreamToken(streamDictionaryToken, data.ToArray());
```

Comment:
can we avoid the `ToArray()` alloc by doing the following:
```csharp
            Memory<byte> data = new byte[dataLength];

            inputBytes.Seek(streamDataStart);
            inputBytes.Read(data.Span);

            inputBytes.Seek(streamDataEnd);

            stream = new StreamToken(streamDictionaryToken, data);
```

Reply:
Thanks, far better, adressed with : https://github.com/UglyToad/PdfPig/pull/838/commits/ea82bf8aa539b87da8edcb3508f1f7a80998ab35